Generate dev and dbg packages

andreas-bok-sociomantic opened the pull request:
Add scripts for generating development and
debug deb packages

andreas-bok-sociomantic committed:
Generate deb and dbg packages

Add scripts for generating development and
debug deb packages

use LIBDRIZZLE_LIBRARY_VERSION variable defined
in configure.ac for naming library in fpm deb
packaging script

Remove unused make target install-deb

ben-palmer-sociomantic commented on deb/build:
```diff
@@ -40,4 +40,22 @@ fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
-	install/=/
+	install/usr/lib/=/usr/lib/
+
+pkgname=libdrizzle-redux-dev
```
I don't really see the difference between this and the `pkgname=libdrizzle-redux` section that includes the header files. Am I missing something?

andreas-bok-sociomantic replied:
AFAIU from Lucas example, the dev version specifies normal version as dependency and only installs headers additionally,

ben-palmer-sociomantic replied:
Right I see.

leandro-lucarella-sociomantic replied:
Not only headers, the regular package should provide only the library with version included: libdrizzle-redux.so.x.y.z (the real file), libdrizzle-redux.so.x.y and libdrizzle-redux.so.x (symlink to the real lib), while the -dev package should provide a symlink without the version (libdrizzle-redux.so). The -dev package should ideally include the static lib too (libdrizzle-redux.a).

See for example:
- http://packages.ubuntu.com/trusty/amd64/libncurses5-dev/filelist / http://packages.ubuntu.com/trusty/amd64/libncurses5/filelist
- http://packages.ubuntu.com/wily/amd64/liblzo2-dev/filelist / http://packages.ubuntu.com/wily/amd64/liblzo2-2/filelist

Also, providing a -dbg package would be quite nice.

ben-palmer-sociomantic commented:
@leandro-lucarella-sociomantic any comments?

andreas-bok-sociomantic commented:
Created scripts for `dbg` and `dev` packages.

leandro-lucarella-sociomantic commented:
If this was tested and works, LGTM :)
Thanks!

andreas-bok-sociomantic commented:
Great, tested by building and installing the `libdrizzle-redux` deb locally and compile the `dive` branch which uses `libdrizzle-redux`. All works.

leandro-lucarella-sociomantic commented:
> tested by building and installing the libdrizzle-redux deb locally and compile the dive branch which uses libdrizzle-redux

I guess you had to install also `libdrizzle-redux-dev` to build, right? Also, you could try with a docker image that the produced binary works when only the `libdrizzle-redux` package is installed (no `-dev` package installed). Making it crash and looking at the backtrace in gdb can help to test if the debug symbols work (you can try with and without the package installed to see the difference).

andreas-bok-sociomantic commented:
Forgot to strip the library before generating the deb packages. 

Tested locally with dbg/dev packages and works like expected 

```diff
--- a/deb/build
+++ b/deb/build
@@ -32,6 +32,8 @@ trap "rm -f '$changelog'; exit 1" INT TERM QUIT
 
 pkgname=libdrizzle-redux
 genchangelog "$pkgname" "$pkgversion" "$pkgmaint" > "$changelog"
+cp install/usr/lib/libdrizzle-redux.so.$libversion ./libdrizzle-redux.so.$libversion
+strip install/usr/lib/libdrizzle-redux.so.$libversion
 fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
        --architecture all \
        --maintainer "$pkgmaint" \
@@ -53,7 +55,7 @@ genchangelog "$pkgname" "$pkgversion" "$pkgmaint" > "$changelog"
 build_id=`readelf -n install/usr/lib/libdrizzle-redux.so.$libversion | \
        sed -n 's/^.*Build ID: \([a-f0-9]\{40\}\).*$/\1/p'`
 debug_file=`printf $build_id | cut -b1-2`/`printf $build_id | cut -b3-`.debug
-objcopy --only-keep-debug install/usr/lib/libdrizzle-redux.so.$libversion libdrizzle-redux.so.$libversion.debug
+objcopy --only-keep-debug ./libdrizzle-redux.so.$libversion libdrizzle-redux.so.$libversion.debug
 fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
        --architecture all \
        --maintainer "$pkgmaint" \
@@ -89,3 +91,5 @@ fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
        install/usr/include/=/usr/include/ \
        install/usr/lib/libdrizzle-redux.so=/usr/lib/ \
        install/usr/lib/libdrizzle-redux.a=/usr/lib/
+
+rm ./libdrizzle-redux.so.$libversion
```

leandro-lucarella-sociomantic commented:
> +rm ./libdrizzle-redux.so.$libversion

You should probably add this to the `trap`, so it's removed even if the user hits Ctrl-C. Or maybe just create it in the `install/` dir?

andreas-bok-sociomantic committed:
Restructure build script

save path to the built library in
a variable so_libname, and refer to
this in the deb build script

andreas-bok-sociomantic commented:
addressed comment by @leandro-lucarella-sociomantic

andreas-bok-sociomantic commented:
rebased on master

ben-palmer-sociomantic commented:
Should the commit title be `Generate dev and dbg packages` rather than `Generate deb and dbg packages`?

andreas-bok-sociomantic commented:
Updated PR title

ben-palmer-sociomantic commented:
LGTM, any other comments @leandro-lucarella-sociomantic ?

leandro-lucarella-sociomantic commented:
Nope, LGTM.

andreas-bok-sociomantic commented:
If no other comments I would appreciate if it was merged :)

ben-palmer-sociomantic commented:
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was 329f056c3fcf92a0e334fa9cf0c35533fe623695, new (rebased) HEAD is dc5c511dc4799516f2b2e1689ab8a33f48b2634b
